### PR TITLE
[AIRFLOW-2062] Add per-connection KMS encryption.

### PR DESCRIPTION
**Pull request by jakahn**
### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AIRFLOW-2062

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
   - For an in-depth explanation of these changes, please see this [design doc](https://docs.google.com/document/d/1qaucGw52aoR96swHQqIYN9nQXn2QKSOWvw6-KyI6Wwc/edit?usp=sharing).
   - Connections can now be encrypted using a local key that is encrypted using a KMS before being stored, instead of the global fernet key. Enabling this behavior requires creating a connection through the CLI (web UI in a coming PR) with the new KMS fields specified.
   - `Connection` and `BaseHook` were also refactored to support loading connections from the `Connection` class (instead of `BaseHook`).

### Tests

- [X] My PR adds the following unit tests:
   - `tests/contrib/hooks/test_gcp_kms_hook.py`
      - `test_encrypt_conn`
      - `test_encrypt_conn`
   - `tests/models.py`
      - `FernetTest`
      - `test_connection_kms_encryption`
      - `test_connection_from_uri_kms_encryption`
      - `test_get_kms_hook_missing`
      - `test_update_kms`

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.

### Code Quality

- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by bolkedebruin** [comment]
This is very tied to a specific implementation. I would rather see the kms_Xx fields be added to the extra options of a connection

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -1785,6 +1792,15 @@ class CLIFactory(object):
             ('--conn_extra',),
             help='Connection `Extra` field, optional when adding a connection',
             type=str),
+        'kms_conn_id': Arg(
+            ('--kms_conn_id',),
+            help='An existing connection to use when encrpting this connection with a '
```
Comment (gerardo) [inline airflow/bin/cli.py]: There's a typo here

**Comment by gerardo** [comment]
@jakahn this PR makes things pretty confusing, as AWS has a product called [KMS](https://aws.amazon.com/kms/) as well, which has been around for longer. The docs you submitted also assume Google Cloud KMS is the _only_ thing called KMS.

I still like the idea, but the implementation is very specific to Google Cloud and hard to reuse for other implementations. Is there a way you could make it more generic?

**Comment by jakahn** [comment]
@bolkedebruin @gerardo, thanks for the feedback.

The intent is that this implementation _should_ be able to support other KMSs in the future, what aspects were you concerned about regarding Amazon KMS integration? The intent is that, for example, an AWS KMS Hook could be added in the future (similar to `GcpKmsHook` now) following the `KmsApiHook` interface (in addition to supporting any AWS-specific features), and then add it to the list of supported KMSs in `get_kms_hook` (`models.py`, Line 883 in this PR). You should then be able to choose between AWS or GCP KMS on a per-connection basis. 

The reason the `kms_*` fields are not stored as part of the `extra` field is so that you can encrypt *any* connection via KMS managed credentials (not just Google connections). Since other connections may not use JSON extras, we didn't want to mess with their extra data.

File: airflow/models.py [hunk 2]
```diff
@@ -122,44 +123,47 @@ class NullFernet(object):
     """
     is_encrypted = False
 
+    def __init__(self, k):
+        LoggingMixin().log.warn(
```
Comment (Fokko) [inline airflow/models.py]: Please add the logging mixin to the class itself: `class NullFernet(LoggingMixin):
`

**Review by Fokko — CHANGES_REQUESTED** [review]

File: airflow/models.py (hunk 2 above)
Comment (jakahn) [reply airflow/models.py]: Fixed!

**Comment by jakahn** [comment]
@bolkedebruin / @Fokko Do you have further feedback?

**Comment by wwlian** [comment]
@bolkedebruin @gerardo @Fokko I understand the concerns that this change might be coupled too tightly to Google Cloud KMS. However, I want to second @jakahn's assurance that this design is agnostic to the key management service being used. 

The only opinionated design included here is that encryption will be performed using the envelope encryption pattern, which is a widely-recognized pattern by [AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#enveloping), [Google Cloud KMS](https://cloud.google.com/kms/docs/envelope-encryption), and [Azure Key Vault](https://docs.microsoft.com/en-us/azure/storage/common/storage-client-side-encryption#encryption-and-decryption-via-the-envelope-technique).

To add to what @jakahn said re: embedding kms_conn_id and kms_extras in the existing _extra column, doing so would create a chicken and egg problem, as their values are needed to decrypt the _extras column.

**Comment by jakahn** [comment]
@Fokko Any further changes?

**Comment by Fokko** [comment]
@jakahn Sorry for the late response, was a bit busy lately. Can you rebase onto master, so I can merge this? Thanks!

**Comment by jakahn** [comment]
@Fokko No problem! Rebased and pushed, Travis running now.

**Comment by gerardo** [comment]
@Fokko @jakahn I feel like this PR should be presented as an [Airflow Improvement Proposal](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals), given the changes being introduced to `models.py`. It should be easy to do, as there is a [design doc](https://docs.google.com/document/d/1qaucGw52aoR96swHQqIYN9nQXn2QKSOWvw6-KyI6Wwc/edit) already.

BTW @jakahn try rebasing your branch now. It's broken at this moment.

**Comment by Fokko** [comment]
@gerardo Good point. I think we should support KMS in the end anyway, but extending the `models.py` is something that we shouldn't do. @jakahn Can you extract the logic from the models.py in a separate file?

**Pull request closed without merging**